File: rs.go
package nrsc5

// Reed-Solomon decoder, ported from Phil Karn's rs_init.c / rs_decode.c
// (GPL), as used by nrsc5 with parameters (8, 0x11d, 1, 1, 8).

const rsA0 = 255 // A0: log(0) = -inf sentinel (== NN for mm=8)

// rsState holds the Galois field tables and generator polynomial.
type rsState struct {
	mm      int    // bits per symbol
	nn      int    // symbols per block ((1<<mm)-1)
	alphaTo []byte // antilog lookup table (index form -> poly form)
	indexOf []byte // log lookup table (poly form -> index form)
	genpoly []byte // generator polynomial (index form)
	nroots  int    // number of parity symbols
	fcr     int    // first consecutive root, index form
	prim    int    // primitive element, index form
	iprim   int    // prim-th root of 1, index form
}

// rsModnn computes x modulo nn using the shift trick from rs_char.h.
func rsModnn(rs *rsState, x uint) uint {
	for x >= uint(rs.nn) {
		x -= uint(rs.nn)
		x = (x >> uint(rs.mm)) + (x & uint(rs.nn))
	}
	return x
}

// initRs initializes a Reed-Solomon codec.
//
//	symsize = symbol size in bits (1-8)
//	gfpoly  = field generator polynomial coefficients
//	fcr     = first root of RS generator polynomial, index form
//	prim    = primitive element to generate polynomial roots, index form
//	nroots  = degree of RS generator polynomial (number of parity symbols)
func initRs(symsize, gfpoly, fcr, prim, nroots uint) *rsState {
	if symsize > 8 {
		return nil
	}
	if fcr >= 1<<symsize {
		return nil
	}
	if prim == 0 || prim >= 1<<symsize {
		return nil
	}
	if nroots >= 1<<symsize {
		return nil
	}

	rs := &rsState{
		mm:     int(symsize),
		nn:     (1 << symsize) - 1,
		nroots: int(nroots),
		fcr:    int(fcr),
		prim:   int(prim),
	}
	rs.alphaTo = make([]byte, rs.nn+1)
	rs.indexOf = make([]byte, rs.nn+1)

	// Generate Galois field lookup tables.
	rs.indexOf[0] = rsA0 // log(zero) = -inf
	rs.alphaTo[rsA0] = 0 // alpha**-inf = 0
	sr := uint(1)
	for i := uint(0); i < uint(rs.nn); i++ {
		rs.indexOf[sr] = byte(i)
		rs.alphaTo[i] = byte(sr)
		sr <<= 1
		if sr&(1<<symsize) != 0 {
			sr ^= gfpoly
		}
		sr &= uint(rs.nn)
	}
	if sr != 1 {
		// field generator polynomial is not primitive
		return nil
	}

	// Find prim-th root of 1, used in decoding.
	var iprim int
	for iprim = 1; iprim%rs.prim != 0; iprim += rs.nn {
	}
	rs.iprim = iprim / rs.prim

	// Form RS code generator polynomial from its roots.
	rs.genpoly = make([]byte, rs.nroots+1)
	rs.genpoly[0] = 1
	root := fcr * prim
	for i := uint(0); i < nroots; i, root = i+1, root+prim {
		rs.genpoly[i+1] = 1

		// Multiply genpoly[] by alpha**(root + x)
		for j := i; j > 0; j-- {
			if rs.genpoly[j] != 0 {
				rs.genpoly[j] = rs.genpoly[j-1] ^
					rs.alphaTo[rsModnn(rs, uint(rs.indexOf[rs.genpoly[j]])+root)]
			} else {
				rs.genpoly[j] = rs.genpoly[j-1]
			}
		}
		// genpoly[0] can never be zero
		rs.genpoly[0] = rs.alphaTo[rsModnn(rs, uint(rs.indexOf[rs.genpoly[0]])+root)]
	}
	// Convert genpoly[] to index form.
	for i := 0; i <= rs.nroots; i++ {
		rs.genpoly[i] = rs.indexOf[rs.genpoly[i]]
	}

	return rs
}

// decodeRs decodes a Reed-Solomon codeword in place. data holds NN symbols
// (for shortened codes, the leading positions must be zero). erasPos holds
// erasure positions (may be nil if noEras == 0). Returns the number of
// corrected symbols, or -1 if the codeword is uncorrectable.
func decodeRs(rs *rsState, data []byte, erasPos []int, noEras int) int {
	var (
		degLambda, el, degOmega int
		i, j, r, k              int
		u, q, tmp, num1, num2   uint
		den, discrR             byte
	)
	nroots := rs.nroots
	nn := rs.nn
	lambda := make([]byte, nroots+1)
	s := make([]byte, nroots)
	b := make([]byte, nroots+1)
	t := make([]byte, nroots+1)
	omega := make([]byte, nroots+1)
	root := make([]uint, nroots)
	reg := make([]byte, nroots+1)
	loc := make([]int, nroots)

	// Form the syndromes; i.e., evaluate data(x) at roots of g(x).
	for i = 0; i < nroots; i++ {
		s[i] = data[0]
	}
	for j = 1; j < nn; j++ {
		for i = 0; i < nroots; i++ {
			if s[i] == 0 {
				s[i] = data[j]
			} else {
				s[i] = data[j] ^ rs.alphaTo[rsModnn(rs, uint(rs.indexOf[s[i]])+uint(rs.fcr+i)*uint(rs.prim))]
			}
		}
	}

	// Convert syndromes to index form, checking for nonzero condition.
	synError := 0
	for i = 0; i < nroots; i++ {
		synError |= int(s[i])
		s[i] = rs.indexOf[s[i]]
	}

	if synError == 0 {
		// Syndromes are zero: data[] is a codeword with no errors.
		return 0
	}

	lambda[0] = 1
	for i := 1; i <= nroots; i++ {
		lambda[i] = 0
	}

	if noEras > 0 {
		// Init lambda to be the erasure locator polynomial.
		lambda[1] = rs.alphaTo[rsModnn(rs, uint(rs.prim*(nn-1-erasPos[0])))]
		for i = 1; i < noEras; i++ {
			u = rsModnn(rs, uint(rs.prim*(nn-1-erasPos[i])))
			for j = i + 1; j > 0; j-- {
				tmp = uint(rs.indexOf[lambda[j-1]])
				if tmp != rsA0 {
					lambda[j] ^= rs.alphaTo[rsModnn(rs, u+tmp)]
				}
			}
		}
	}
	for i = 0; i <= nroots; i++ {
		b[i] = rs.indexOf[lambda[i]]
	}

	// Berlekamp-Massey algorithm to determine error+erasure locator poly.
	r = noEras
	el = noEras
	for {
		r++
		if uint(r) > uint(nroots) {
			break
		}
		// Compute discrepancy at the r-th step in poly form.
		discrR = 0
		for i = 0; i < r; i++ {
			if lambda[i] != 0 && s[r-i-1] != rsA0 {
				discrR ^= rs.alphaTo[rsModnn(rs, uint(rs.indexOf[lambda[i]])+uint(s[r-i-1]))]
			}
		}
		discrR = rs.indexOf[discrR] // index form
		if discrR == rsA0 {
			// B(x) <-- x*B(x)
			copy(b[1:], b[:nroots])
			b[0] = rsA0
		} else {
			// T(x) <-- lambda(x) - discr_r*x*b(x)
			t[0] = lambda[0]
			for i = 0; i < nroots; i++ {
				if b[i] != rsA0 {
					t[i+1] = lambda[i+1] ^ rs.alphaTo[rsModnn(rs, uint(discrR)+uint(b[i]))]
				} else {
					t[i+1] = lambda[i+1]
				}
			}
			if 2*el <= r+noEras-1 {
				el = r + noEras - el
				// B(x) <-- inv(discr_r) * lambda(x)
				for i = 0; i <= nroots; i++ {
					if lambda[i] == 0 {
						b[i] = rsA0
					} else {
						b[i] = byte(rsModnn(rs, uint(int(rs.indexOf[lambda[i]])-int(discrR)+nn)))
					}
				}
			} else {
				// B(x) <-- x*B(x)
				copy(b[1:], b[:nroots])
				b[0] = rsA0
			}
			copy(lambda[:nroots+1], t[:nroots+1])
		}
	}

	// Convert lambda to index form and compute deg(lambda(x)).
	degLambda = 0
	for i = 0; i <= nroots; i++ {
		lambda[i] = rs.indexOf[lambda[i]]
		if lambda[i] != rsA0 {
			degLambda = i
		}
	}
	// Find roots of the error+erasure locator polynomial by Chien search.
	copy(reg[1:], lambda[1:1+nroots])
	count := 0
	for i, k = 1, rs.iprim-1; i <= nn; i, k = i+1, int(rsModnn(rs, uint(k+rs.iprim))) {
		q = 1 // lambda[0] is always 1 (index form A0 handling below)
		for j = degLambda; j > 0; j-- {
			if reg[j] != rsA0 {
				reg[j] = byte(rsModnn(rs, uint(reg[j])+uint(j)))
				q ^= uint(rs.alphaTo[reg[j]])
			}
		}
		if q != 0 {
			continue // not a root
		}
		// Store root (index form) and error location number.
		root[count] = uint(i)
		loc[count] = k
		count++
		if count == degLambda {
			break
		}
	}
	if degLambda != count {
		// deg(lambda) unequal to number of roots => uncorrectable
		return -1
	}

	// Compute err+eras evaluator poly omega(x) = s(x)*lambda(x)
	// (modulo x**NROOTS), in index form. Also find deg(omega).
	degOmega = 0
	for i = 0; i < nroots; i++ {
		tmp = 0
		j = degLambda
		if j > i {
			j = i
		}
		for ; j >= 0; j-- {
			if s[i-j] != rsA0 && lambda[j] != rsA0 {
				tmp ^= uint(rs.alphaTo[rsModnn(rs, uint(s[i-j])+uint(lambda[j]))])
			}
		}
		if tmp != 0 {
			degOmega = i
		}
		omega[i] = rs.indexOf[byte(tmp)]
	}
	omega[nroots] = rsA0

	// Compute error values in poly form:
	//   num1 = omega(inv(X(l))), num2 = inv(X(l))**(FCR-1),
	//   den = lambda_pr(inv(X(l))) — all in poly form.
	for j = count - 1; j >= 0; j-- {
		num1 = 0
		for i = degOmega; i >= 0; i-- {
			if omega[i] != rsA0 {
				num1 ^= uint(rs.alphaTo[rsModnn(rs, uint(omega[i])+uint(i)*root[j])])
			}
		}
		num2 = uint(rs.alphaTo[rsModnn(rs, root[j]*uint(rs.fcr-1)+uint(nn))])
		den = 0

		// lambda[i+1] for i even is the formal derivative lambda_pr.
		for i = minInt(uint(degLambda), uint(nroots-1)) &^ 1; i >= 0; i -= 2 {
			if lambda[i+1] != rsA0 {
				den ^= rs.alphaTo[rsModnn(rs, uint(lambda[i+1])+uint(i)*root[j])]
			}
		}
		if den == 0 {
			return -1
		}
		// Apply error to data.
		if num1 != 0 {
			data[loc[j]] ^= rs.alphaTo[rsModnn(rs, uint(rs.indexOf[byte(num1)])+
				uint(rs.indexOf[byte(num2)])+uint(nn)-uint(rs.indexOf[den]))]
		}
	}

	if erasPos != nil {
		for i = 0; i < count; i++ {
			erasPos[i] = loc[i]
		}
	}
	return count
}

func minInt(a, b uint) int {
	if a < b {
		return int(a)
	}
	return int(b)
}

// encodeRs computes the NROOTS parity symbols for a systematic codeword
// (Karn-style). data holds NN-NROOTS message symbols; the parity is
// written to parity. This is provided for testing/verification.
func encodeRs(rs *rsState, data []byte, parity []byte) {
	nroots := rs.nroots
	for i := 0; i < nroots; i++ {
		parity[i] = 0
	}
	for i := 0; i < rs.nn-nroots; i++ {
		feedback := uint(rs.indexOf[data[i]^parity[0]])
		if feedback != rsA0 {
			for j := 1; j < nroots; j++ {
				parity[j] ^= rs.alphaTo[rsModnn(rs, feedback+uint(rs.genpoly[nroots-j]))]
			}
			copy(parity, parity[1:])
			parity[nroots-1] = rs.alphaTo[rsModnn(rs, feedback+uint(rs.genpoly[0]))]
		} else {
			copy(parity, parity[1:])
			parity[nroots-1] = 0
		}
	}
}

File: python/nrsc5py/rs.py
"""Reed-Solomon decoder (mirrors rs.go — a port of Phil Karn's
decode_rs_char, as used by nrsc5 for the L2 frame headers).

nrsc5 uses RS(255,247) shortened: 8 parity symbols over GF(256)
(poly 0x11d, fcr=1, prim=1). The 96-byte header is placed at the END
of a virtual 255-symbol block whose first 159 symbols are zero.

Syndromes, Berlekamp-Massey, Chien search and Forney's algorithm are
JIT-compiled; the Galois field tables are built at init.
"""

from __future__ import annotations

import numpy as np

try:
    from numba import njit
except ImportError:  # pragma: no cover
    def njit(*args, **kwargs):
        if len(args) == 1 and callable(args[0]):
            return args[0]
        def wrap(fn):
            return fn
        return wrap


class RsState:
    """Galois field tables + generator polynomial (struct rs)."""

    def __init__(self, symsize: int = 8, gfpoly: int = 0x11D,
                 fcr: int = 1, prim: int = 1, nroots: int = 8):
        self.mm = symsize
        self.nn = (1 << symsize) - 1
        self.nroots = nroots
        self.fcr = fcr
        self.prim = prim
        self.a0 = self.nn  # A0 sentinel: log(0) = -inf

        self.alpha_to = np.zeros(self.nn + 1, dtype=np.uint8)
        self.index_of = np.zeros(self.nn + 1, dtype=np.uint8)
        self.index_of[0] = self.a0
        self.alpha_to[self.a0] = 0
        sr = 1
        for i in range(self.nn):
            self.index_of[sr] = i
            self.alpha_to[i] = sr
            sr <<= 1
            if sr & (1 << symsize):
                sr ^= gfpoly
            sr &= self.nn
        if sr != 1:
            raise ValueError("field generator polynomial is not primitive")

        # prim-th root of 1 (iprim), used in the Chien search.
        iprim = 1
        while iprim % self.prim != 0:
            iprim += self.nn
        self.iprim = iprim // self.prim

        # Generator polynomial from its roots, in index form.
        self.genpoly = np.zeros(nroots + 1, dtype=np.uint8)
        self.genpoly[0] = 1
        root = fcr * prim
        for i in range(nroots):
            self.genpoly[i + 1] = 1
            for j in range(i, 0, -1):
                if self.genpoly[j] != 0:
                    self.genpoly[j] = self.genpoly[j-1] ^ self.alpha_to[
                        _modnn(self, int(self.index_of[self.genpoly[j]]) + root)]
                else:
                    self.genpoly[j] = self.genpoly[j-1]
            self.genpoly[0] = self.alpha_to[
                _modnn(self, int(self.index_of[self.genpoly[0]]) + root)]
            root += prim
        self.genpoly = np.array(
            [self.index_of[v] for v in self.genpoly], dtype=np.uint8)


def _modnn(rs: RsState, x: int) -> int:
    while x >= rs.nn:
        x -= rs.nn
        x = (x >> rs.mm) + (x & rs.nn)
    return x


def init_rs(symsize=8, gfpoly=0x11D, fcr=1, prim=1, nroots=8) -> RsState:
    return RsState(symsize, gfpoly, fcr, prim, nroots)


@njit(cache=True)
def _decode_rs_core(data, alpha_to, index_of, genpoly_unused,
                    nn, nroots, fcr, prim, iprim, a0, eras_pos, no_eras):
    """Karn's decode_rs_char, JIT-compiled. Returns corrections (-1 = fail).

    All polynomial arrays work in int32 (the C version's uint8 arithmetic
    relies on wrap-around in a few places; Karn's code keeps values < 256
    so int32 is safe and numba-friendly).
    """
    deg_lambda = 0
    el = 0
    deg_omega = 0
    i = 0
    j = 0
    r = 0
    k = 0
    lambda_ = np.zeros(nroots + 1, dtype=np.int32)
    s = np.zeros(nroots, dtype=np.int32)
    b = np.zeros(nroots + 1, dtype=np.int32)
    t = np.zeros(nroots + 1, dtype=np.int32)
    omega = np.zeros(nroots + 1, dtype=np.int32)
    root = np.zeros(nroots, dtype=np.int64)
    reg = np.zeros(nroots + 1, dtype=np.int32)
    loc = np.zeros(nroots, dtype=np.int64)

    # Form the syndromes: evaluate data(x) at the roots of g(x).
    for i in range(nroots):
        s[i] = data[0]
    for j in range(1, nn):
        for i in range(nroots):
            if s[i] == 0:
                s[i] = data[j]
            else:
                x = index_of[s[i]] + (fcr + i) * prim
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                s[i] = data[j] ^ alpha_to[x]

    # Convert syndromes to index form, checking for nonzero condition.
    syn_error = 0
    for i in range(nroots):
        syn_error |= s[i]
        s[i] = index_of[s[i]]

    if syn_error == 0:
        return 0

    lambda_[0] = 1
    for i in range(1, nroots + 1):
        lambda_[i] = 0

    if no_eras > 0:
        # Init lambda to the erasure locator polynomial.
        x = prim * (nn - 1 - eras_pos[0])
        while x >= nn:
            x -= nn
            x = (x >> 8) + (x & nn)
        lambda_[1] = alpha_to[x]
        for i in range(1, no_eras):
            x = prim * (nn - 1 - eras_pos[i])
            while x >= nn:
                x -= nn
                x = (x >> 8) + (x & nn)
            u = x
            for j in range(i + 1, 0, -1):
                tmp = index_of[lambda_[j - 1]]
                if tmp != a0:
                    x2 = u + tmp
                    while x2 >= nn:
                        x2 -= nn
                        x2 = (x2 >> 8) + (x2 & nn)
                    lambda_[j] ^= alpha_to[x2]

    for i in range(nroots + 1):
        b[i] = index_of[lambda_[i]]

    # Berlekamp-Massey.
    r = no_eras
    el = no_eras
    while True:
        r += 1
        if r > nroots:
            break
        discr_r = 0
        for i in range(r):
            if lambda_[i] != 0 and s[r - i - 1] != a0:
                x = index_of[lambda_[i]] + s[r - i - 1]
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                discr_r ^= alpha_to[x]
        discr_r = index_of[discr_r]
        if discr_r == a0:
            for j in range(nroots, 0, -1):
                b[j] = b[j - 1]
            b[0] = a0
        else:
            t[0] = lambda_[0]
            for i in range(nroots):
                if b[i] != a0:
                    x = discr_r + b[i]
                    while x >= nn:
                        x -= nn
                        x = (x >> 8) + (x & nn)
                    t[i + 1] = lambda_[i + 1] ^ alpha_to[x]
                else:
                    t[i + 1] = lambda_[i + 1]
            if 2 * el <= r + no_eras - 1:
                el = r + no_eras - el
                for i in range(nroots + 1):
                    if lambda_[i] == 0:
                        b[i] = a0
                    else:
                        x = index_of[lambda_[i]] - discr_r + nn
                        while x >= nn:
                            x -= nn
                            x = (x >> 8) + (x & nn)
                        b[i] = x
            else:
                for j in range(nroots, 0, -1):
                    b[j] = b[j - 1]
                b[0] = a0
            for i in range(nroots + 1):
                lambda_[i] = t[i]

    # Convert lambda to index form and compute deg(lambda).
    deg_lambda = 0
    for i in range(nroots + 1):
        lambda_[i] = index_of[lambda_[i]]
        if lambda_[i] != a0:
            deg_lambda = i

    # Chien search for roots of the locator polynomial.
    for i in range(1, nroots + 1):
        reg[i] = lambda_[i]
    count = 0
    k = iprim - 1
    i = 1
    while i <= nn:
        q = 1
        for j in range(deg_lambda, 0, -1):
            if reg[j] != a0:
                x = reg[j] + j
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                reg[j] = x
                q ^= alpha_to[reg[j]]
        if q == 0:
            root[count] = i
            loc[count] = k
            count += 1
            if count == deg_lambda:
                break
        x = k + iprim
        while x >= nn:
            x -= nn
            x = (x >> 8) + (x & nn)
        k = x
        i += 1
    if deg_lambda != count:
        return -1

    # Omega = s(x)*lambda(x) mod x**nroots.
    deg_omega = 0
    for i in range(nroots):
        tmp = 0
        j = deg_lambda if deg_lambda < i else i
        while j >= 0:
            if s[i - j] != a0 and lambda_[j] != a0:
                x = s[i - j] + lambda_[j]
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                tmp ^= alpha_to[x]
            j -= 1
        if tmp != 0:
            deg_omega = i
        omega[i] = index_of[tmp]
    omega[nroots] = a0

    # Forney's algorithm for the error values.
    for j in range(count - 1, -1, -1):
        num1 = 0
        for i in range(deg_omega, -1, -1):
            if omega[i] != a0:
                x = omega[i] + i * root[j]
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                num1 ^= alpha_to[x]
        x = root[j] * (fcr - 1) + nn
        while x >= nn:
            x -= nn
            x = (x >> 8) + (x & nn)
        num2 = alpha_to[x]
        den = 0
        lim = deg_lambda if deg_lambda < (nroots - 1) else (nroots - 1)
        i = lim & ~1
        while i >= 0:
            if lambda_[i + 1] != a0:
                x = lambda_[i + 1] + i * root[j]
                while x >= nn:
                    x -= nn
                    x = (x >> 8) + (x & nn)
                den ^= alpha_to[x]
            i -= 2
        if den == 0:
            return -1
        if num1 != 0:
            x = index_of[num1] + index_of[num2] + nn - index_of[den]
            while x >= nn:
                x -= nn
                x = (x >> 8) + (x & nn)
            data[loc[j]] ^= alpha_to[x]

    return count


def decode_rs(rs: RsState, data: np.ndarray, eras_pos=None, no_eras: int = 0) -> int:
    """Decode a codeword in place; returns corrections or -1."""
    return _decode_rs_core(
        data, rs.alpha_to, rs.index_of, rs.genpoly,
        rs.nn, rs.nroots, rs.fcr, rs.prim, rs.iprim, rs.a0,
        eras_pos if eras_pos is not None else np.zeros(0, dtype=np.int64),
        no_eras)
